homeassistant/components/p1_monitor/__init__.py

```diff
@@ -0,0 +1,91 @@
+"""The P1 Monitor integration."""
+from __future__ import annotations
+
+from datetime import timedelta
+
+from p1monitor import P1Monitor, Phases, Settings, SmartMeter
+
+from homeassistant.components.sensor import DOMAIN as SENSOR_DOMAIN
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import (
+    CONF_TIME_BETWEEN_UPDATE,
+    DEFAULT_TIME_BETWEEN_UPDATE,
+    DOMAIN,
+    LOGGER,
+    SERVICE_PHASES,
+    SERVICE_SETTINGS,
+    SERVICE_SMARTMETER,
+)
+
+PLATFORMS = (SENSOR_DOMAIN,)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up P1 Monitor from a config entry."""
+    hass.data.setdefault(DOMAIN, {}).setdefault(entry.entry_id, {})
+
+    client = P1Monitor(host=entry.data[CONF_HOST])
+
+    min_time_between_updates = timedelta(
+        seconds=entry.options.get(CONF_TIME_BETWEEN_UPDATE, DEFAULT_TIME_BETWEEN_UPDATE)
+    )
+
+    async def update_smartmeter() -> SmartMeter:
+        return await client.smartmeter()
+
+    smartmeter: DataUpdateCoordinator[SmartMeter] = DataUpdateCoordinator(
+        hass,
+        LOGGER,
+        name=f"{DOMAIN}_{SERVICE_SMARTMETER}",
+        update_interval=min_time_between_updates,
+        update_method=update_smartmeter,
+    )
+    await smartmeter.async_config_entry_first_refresh()
+    hass.data[DOMAIN][entry.entry_id][SERVICE_SMARTMETER] = smartmeter
+
+    async def update_phases() -> Phases:
+        return await client.phases()
+
+    phases: DataUpdateCoordinator[Phases] = DataUpdateCoordinator(
+        hass,
+        LOGGER,
+        name=f"{DOMAIN}_{SERVICE_PHASES}",
+        update_interval=min_time_between_updates,
+        update_method=update_phases,
+    )
+    await phases.async_config_entry_first_refresh()
+    hass.data[DOMAIN][entry.entry_id][SERVICE_PHASES] = phases
+
+    async def update_settings() -> Settings:
+        return await client.settings()
+
+    settings: DataUpdateCoordinator[Settings] = DataUpdateCoordinator(
+        hass,
+        LOGGER,
+        name=f"{DOMAIN}_{SERVICE_SETTINGS}",
+        update_interval=min_time_between_updates,
+        update_method=update_settings,
+    )
+    await settings.async_config_entry_first_refresh()
+    hass.data[DOMAIN][entry.entry_id][SERVICE_SETTINGS] = settings
```

Comment:
Is there any real benefit to using separate update coordinators. Feels like just more overhead for hass. I suggest just have a single update coordinator that grabs all the data.

Reply:
any suggestions of how I can put this in a single update coordinator?

Reply:
Let the update_method return a tuple or dict of values.

Reply:
Yeah, we only need different coordinators if we need different update intervals or need to avoid fetching some data depending on a setting.

Reply:
Trying to figure out how to make this

Reply:
Do you have a code suggestion for me? Because my previous attempts to solve this failed.

Reply:
```
async def _coordinator_update()
  data = {}
  data["1"] = await client.getdata1()
  data["2"] = await client.getdata2()
  return data
```

Then in your entities.
```
data1 = self.coordinator.data["1"]
```

Reply:
Alright with thnx to @MartinHjelmare, we were able to convert it to a single DataUpdateCoordinator.